src/uct/ib/base/ib_iface.c

```diff
@@ -548,8 +549,25 @@ ucs_status_t uct_ib_iface_get_device_address(uct_iface_h tl_iface,
     return UCS_OK;
 }
 
+static int uct_ib_iface_roce_is_reachable_via_routing(unsigned subnet_mask,
+        const uct_ib_device_gid_info_t *local_gid_info,
+        const uct_ib_address_t *remote_ib_addr)
+{
+    size_t size;
+
+    if (!uct_ib_device_is_potentially_reachable(subnet_mask, local_gid_info)) {
+        return 0;
+    }
+
+    ucs_address_family_sizeof_ip(local_gid_info->roce_info.addr_family, &size);
+    size = sizeof(local_gid_info->gid) + (subnet_mask >> 3) - size;
+
+    return (0 == memcmp(&local_gid_info->gid, &remote_ib_addr->flags + 1, size));
```

Comment:
the IPv4 address is at the end of the gid
need to reuse a generic function which converts GID to IPv4/6, there should be such function in UD transport for filtering incoming packets

Reply:
ok

Reply:
to be clear - do you mean `uct_ud_grh_get_dgid`? today it takes the IP from the GRH and generates a dgid, so it's not very similar to what I need (to read an IP from a gid). Do you prefer I move some of this function to `uct/ib/base` and make UD use it?

Reply:
yes, need generic func in ib/base to convert gid->ip